ingest/bin/post_process_metadata.py

```diff
@@ -0,0 +1,91 @@
+#! /usr/bin/env python3
+
+"""Reformat pandas DataTables for a pathogen build.
+
+Expecting one argument, a NCBI Virus metadata.tsv file
+"""
+# ===== Dependencies
+import argparse
+import os
+import sys
+
+import numpy as np
+import pandas as pd
+
+
+def parse_args():
+  # Main help command
+  parser = argparse.ArgumentParser(
+    description = "USAGE: python post_process_metadata.py --metadata 'data/metadata.tsv'"
+  )
+  # Add first argument
+  parser.add_argument(
+    "--metadata",
+    help = "NCBI Virus metadata.tsv file.",
+    required = True
+  )
+  parser.add_argument(
+    "--outfile", default="processed_metadata.tsv",
+    help = "Processed metdata file name. [default: processed_metadata.tsv] ",
+    required = False
+  )
+  
+  return parser.parse_args()
+
+# === Private methods
+
+def _set_strain_name(record):
+  """Check Isolate_s and Strain_s to find the strain name"""
+  strain_name=""
+  if record['strain'] != record['accession']:
+    strain_name=record['strain']
+  elif record['strain'] == record['accession'] and pd.notna(record['strain_s']):
+    strain_name=record['strain_s']
+  else:
+    strain_name=record['strain']
+
+  return strain_name.replace(' ', '_').replace('-',"_").replace(".","_").replace("(","_").replace(")","_")
```

Comment:
This would be a great place for either [Python's built-in `re.sub` function](https://docs.python.org/3/library/re.html#re.sub) or the regular expression functionality of [the pandas Series method `replace`](https://pandas.pydata.org/docs/reference/api/pandas.Series.replace.html). Since you are already working with pandas Series objects here, you could replace this line with:

```python
return strain_name.replace(r"[ -.,)]", "_", regex=True)
```

Reply:
While I like an elegant regex, I'm going to push back on this to avoid escaping special characters like `.` matching everything, `()` capturing substrings, `-` defining a range of values. Also due to:

* https://stackoverflow.com/questions/5668947/use-pythons-string-replace-vs-re-sub

Reply:
Good point about escaping! I hadn't tested the regex above, but you're right that those special characters should be escaped like so:

```python
>>> import re
>>> s = "A/(This is a strain . -with-extra stuff)/2020"
>>> re.sub(r"[ \-\.\(\)]", "_", s)
'A/_This_is_a_strain____with_extra_stuff_/2020'
```

I see the arguments for being explicit and fast from the SO posts, although speed is less of an issue than readability and maintainability in this context. I was hoping that the regex would be simpler and easier to maintain than a chain of replace calls, but I can see that the actual regex you'd need is probably not much clearer.

Reply:
`str.replace()` is fine by me, but I wanted to follow up to clarify about regexes.

Metachars aren't interpreted as such inside character classes `[…]`, so the escaping isn't necessary. `[ -.()]` works just fine. I often escape `.` with the expression `[.]` which I find more readable in many contexts than `\.`.

One caveat here: the hyphen should be in the first or last position so it's not interpreted as defining a character range (e.g. ` ` to `.` in the ASCII char table, or ` !"#$%&'*+,-.`).

```python
>>> re.sub(r"[ -.()]", "_", """ -.()!"#$%&'*+, """)
'________________'

>>> re.sub(r"[ .()-]", "_", """ -.()!"#$%&'*+, """)
'_____!"#$%&\'*+,_'
```